_posts/2024-05-03-dynamic-client-ssl-context.adoc

```diff
@@ -0,0 +1,128 @@
+If the destination does not match any of the rules, the configured default SSLContext will be used. If it is not configured, then `SSLContext.getDefault()` will be used.
+
+It is also possible to configure this dynamic SSL context to be a `default SSL context` on the server:
+
+```
+/subsystem=elytron:write-attribute(name=default-ssl-context,value=dynamicClientSSLContext)
+reload
+```
+
+If this is the case, standalone clients in a deployed applications that use the `SSLContext.getDefault()` will use this dynamic SSL context for their requests.
+
+== Example use on the client side
+
+To use the dynamic client SSL context with the standalone clients, make sure you have the elytron module `wildfly-elytron-dynamic-ssl` on your classpath. You can then instantiate the `org.wildfly.security.dynamic.ssl.DynamicSSLContext` as shown below:
+
+```
+DynamicSSLContext dynamicSSLContext = new DynamicSSLContext();
+```
+
+The DynamicSSLContext instance loads the authentication context from the classpath as is described in the https://docs.wildfly.org/32/WildFly_Elytron_Security.html#the-default-configuration-approach[elytron client documentation].
+
+You can also pass an existing `AuthenticationContext` instance to the `org.wildfly.security.dynamic.ssl.DynamicSSLContextImpl` class and use it directly with the DynamicSSLContext: `new DynamicSSLContext(new DynamicSSLContextImpl(myAuthenticationContextInstance));`
+
+Same as on the server side, the client configuration can also specify different rules for different SSL contexts. These rules are used by the DynamicSSLContext to select an SSLContext to delegate to.
+
+As always, you can use either the programmatic approach to provide the authentication context or the configuration file approach. Below is an example of SSL context rules in the configuration file:
+
+```
+<ssl-context-rules>
+    <rule use-ssl-context="client-context1">
+        <match-host name="www.example.org"/>
+        <match-port number="443"/>
+    </rule>
+    <rule use-ssl-context="client-context2">
+        <match-host name="localhost"/>
+        <match-port number="8443"/>
+    </rule>
+    <rule use-ssl-context="client-context3">
+        <match-port number="9443"/>
+    </rule>
+        <rule use-ssl-context="default-context">
+    </rule>
+</ssl-context-rules>
+```
+
+Note: Since the `org.wildfly.security.dynamic.ssl.DynamicSSLContextSPI` is an SPI, you can use your own implementation to provide SSL context selection instead of relying on the elytron client.
+
+== Example using reverse proxies
+
+We will use a simple example consisting of a https://github.com/wildfly-security-incubator/elytron-examples/blob/main/dynamic-ssl-reverse-proxies/configure.cli[configure.cli] file. This file configures 2 ports secured with TLS and 2 reverse proxies connecting to these ports.
```

Comment:
Let me know if I should copy relevant commands from configure.cli file here or if it is okay to just reference it

Reply:
I think it's ok to just reference it and just keep this as a blog post (as opposed to a use-case driven guide).

For a use-case driven guide, it would be good to go through the commands in order to show the steps required to accomplish the task.

The current format of the post is fine though, no need to update it.